src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -907,22 +907,72 @@
        (fn [[col-names & rows]]
          (into existing-path-ids rows))))))
 
-(defn insert-facts! [factset-id pids valuemaps]
-  (when (seq valuemaps)
-    (let [vm-keys (remove #{:large_value_hash} (keys (first valuemaps)))
-          cols (concat [:factset_id :fact_path_id :large_value_hash]
-                       vm-keys)]
+(def ^:const db-fv-string 0)
+(def ^:const db-fv-int 1)
+(def ^:const db-fv-float 2)
+(def ^:const db-fv-bool 3)
+(def ^:const db-fv-nil 4)
+(def ^:const db-fv-struct 5)
+
+(defn fact-value-type-id [x]
+  (cond
+    (keyword? x) db-fv-string
+    (string? x) db-fv-string
+    (integer? x) db-fv-int
+    (float? x) db-fv-float
+    (kitchensink/boolean? x) db-fv-bool
+    (nil? x) db-fv-nil
+    (coll? x) db-fv-struct
+    :else (throw (Exception. (trs "Fact value {0} has unexpected type"
+                                  (pr-str x))))))
+
+(def ^:const value-row-cols
+  ;; Must match order in value->value-row immediately below
+  [:value_type_id
+   :large_value_hash
+   :value
+   :value_string
+   :value_integer
+   :value_float
+   :value_boolean])
+
+(defn storage-hash [x]
+  ;; If the return type changes here, update the PGobject check in update-facts.
+  (-> x shash/generic-identity-hash sutils/munge-hash-for-storage))
+
+(defn value->value-row
+  ([value] (value->value-row value nil))
+  ([value hash]
+   (let [jsonb (when-not (nil? value)
+                 (sutils/munge-jsonb-for-storage value))
+         large? (and jsonb (>= (count (.getValue jsonb))
+                               facts/large-value-threshold))
+         hash (when large?
+                (or hash (storage-hash value)))]
+     (value->value-row value hash jsonb)))
+  ([value hash jsonb]
+   (let [id (fact-value-type-id value)]
+     (cond
```

Comment:
I was thinking about how to improve this. You could let bind `(let [v [(fact-value-type-id value) hash jsonb]]`. Below then becomes `(into v [value nil nil nil])` etc.

I also thought about case here, I thought it might work since the vars are const, but it won't condp would reduce the repeating yourself there, but there's a hit to clarity too.